cou/apps/app.py

```diff
@@ -245,24 +224,121 @@ def __str__(self) -> str:
+        workload version.
+        :return: The latest compatible OpenStack release.
+        :rtype: OpenStackRelease
+        """
+        if self.is_os_channel_based:
+            return self.channel_codename
+
+        try:
+            return max(
+                OpenStackCodenameLookup.find_compatible_versions(self.charm, unit.workload_version)
+            )
+        except ValueError as exc:
+            raise ApplicationError(
+                f"'{self.name}' with workload version {unit.workload_version} has no "
+                f"compatible OpenStack release."
+            ) from exc
+
+    @property
+    def channel(self) -> str:
+        """Get charm channel of the application.
+
+        :return: Charm channel. E.g: ussuri/stable
+        :rtype: str
+        """
+        return self._channel
+
+    @channel.setter
+    def channel(self, charm_channel: str) -> None:
+        """Set charm channel of the application.
+
+        When application comes from charm store, the channel won't be OpenStack related.
+        :param charm_channel: Charm channel. E.g: ussuri/stable
+        :type charm_channel: str
+        :raises ValueError:  Exception raised when channel is not a valid OpenStack
+            channel.
+        """
+        if (
+            self.is_from_charm_store
+            or (not self.is_os_channel_based and self.is_subordinate)
+            or self.is_valid_track
+        ):
+            self._channel = charm_channel
+            return
+        raise ValueError(self.channel_err_msg(charm_channel))
+
+    def channel_err_msg(self, charm_channel: str) -> str:
+        """Error message when channel is not valid.
+
+        :param charm_channel: Charm channel.
+        :type charm_channel: str
+        :return: error message
+        :rtype: str
+        """
+        return f"'{self.name}' has invalid channel: '{charm_channel}'"
+
     @property
     def possible_current_channels(self) -> list[str]:
         """Return the possible current channels based on the current OpenStack release.
 
         :return: The possible current channels for the application. E.g: ["ussuri/stable"]
         :rtype: list[str]
         """
-        return [f"{self.current_os_release.codename}/stable"]
+        return self._get_channels_based_on_os(self.current_os_release)
 
     def target_channel(self, target: OpenStackRelease) -> str:
         """Return the appropriate channel for the passed OpenStack target.
 
         :param target: OpenStack release as target to upgrade.
         :type target: OpenStackRelease
+        :raises ApplicationError: When cannot find a suitable target channel.
         :return: The next channel for the application. E.g: victoria/stable
         :rtype: str
         """
-        return f"{target.codename}/stable"
+        targets_channels = self._get_channels_based_on_os(target)
+        if targets_channels:
+            return targets_channels[-1]
```

Comment:
To my this part does not make since default `_get_channels_based_on_os` return list with one element, I know that in `OpenStackAuxiliaryApplication` you overwritten it, but why not override this function directly.